bleak/backends/bluezdbus/client.py

```diff
@@ -214,50 +416,61 @@ async def _cleanup_dbus_resources(self) -> None:
+        Free all the allocated resource in DBus. Use this method to
         eventually cleanup all otherwise leaked resources.
         """
-        self._char_path_to_handle.clear()
-        await self._cleanup_notifications()
-        await self._cleanup_dbus_resources()
+        await self._remove_signal_handlers()
+        self._disconnect_message_bus()
 
     async def disconnect(self) -> bool:
         """Disconnect from the specified GATT server.
 
         Returns:
             Boolean representing if device is disconnected.
 
+        Raises:
+            BleakDBusError: If there was a D-Bus error
+            asyncio.TimeoutError if the device was not disconnected within 10 seconds
         """
-        logger.debug("Disconnecting from BLE device...")
+        logger.debug(f"Disconnecting ({self._device_path})")
+
         if self._bus is None:
             # No connection exists. Either one hasn't been created or
-            # we have already called disconnect and closed the txdbus
+            # we have already called disconnect and closed the D-Bus
             # connection.
+            logger.debug(f"already disconnected ({self._device_path})")
             return True
 
-        # Remove all residual notifications.
-        await self._cleanup_notifications()
-
-        # Try to disconnect the actual device/peripheral
-        try:
-            await self._bus.callRemote(
-                self._device_path,
-                "Disconnect",
-                interface=defs.DEVICE_INTERFACE,
-                destination=defs.BLUEZ_SERVICE,
-            ).asFuture(asyncio.get_event_loop())
-        except Exception as e:
-            logger.error("Attempt to disconnect device failed: {0}".format(e))
-
-        is_disconnected = not await self.is_connected()
+        if self._disconnecting_event:
+            # another call to disconnect() is already in progress
+            logger.debug(f"already in progress ({self._device_path})")
+            await asyncio.wait_for(self._disconnecting_event.wait(), timeout=10)
+        elif self.is_connected:
+            self._disconnecting_event = asyncio.Event()
+            try:
+                # Try to disconnect the actual device/peripheral
+                reply = await self._bus.call(
+                    Message(
+                        destination=defs.BLUEZ_SERVICE,
+                        path=self._device_path,
+                        interface=defs.DEVICE_INTERFACE,
+                        member="Disconnect",
+                    )
+                )
+                assert_reply(reply)
+                await asyncio.wait_for(self._disconnecting_event.wait(), timeout=10)
+            finally:
+                self._disconnecting_event = None
 
-        await self._cleanup_dbus_resources()
+        # sanity check to make sure _cleanup_all() was triggered by the
+        # "PropertiesChanged" signal handler and that it completed successfully
+        assert self._bus is None
```

Comment:
`dbus-next` uses `assert` quite a bit, so it was rubbing off on me. I can look at removing this since we don't really use `assert` elsewhere in Bleak, but the places I did use it are cases where "this should theoretically never happen"